### ISSUE TITLE
Chapter 6.3 - why shuffle training data for time-series RNN?

### ISSUE THREAD

**Issue by neilgd** [fchollet/deep-learning-with-python-notebooks#66]
In listing 6.33, for building a RNN for the jena weather dataset, the 'generator' function has a 'shuffle' parameter:

```
if shuffle:
            rows = np.random.randint(
                min_index + lookback, max_index, size=batch_size)
```
I don't understand this - given it's time data, why would you want to randomize its order? Can anybody explain?

Thanks!

**Comment by hcgaron** [comment]
@neilgd I believe the reason we have a `shuffle` parameter is because the time series is not stationary, so contiguous data is likely to be highly correlated.  I think the confusion here is that `rows` is simply returning a list of indices that we will use later to return samples.  The samples will still be five days worth of contiguous data with a corresponding temperature target 24 hours into the future.

I think what I wrote above isn't entirely clear, so let me try to explain the relevant parts of the code:

`        if shuffle:
            rows = np.random.randint(min_index + lookback, max_index, size=batch_size)`

^^This line of code simply gets a bunch of numbers at random between `min_index + lookback` and `max_index`.  These aren't actually the samples yet... they are just single numbers we will use to build our samples in a later line of code.

`            
        samples = np.zeros((len(rows), lookback // step, data.shape[-1]))`
`        targets = np.zeros((len(rows),))  # corresponding targets for the samples`

^^ These two lines of code set up the dimensions of the tensors that will hold our actual samples and targets.  `samples` is a 3D tensor with shape `(batch_size, steps, features)`.  `targets` is a tensor with one prediction for each sample in the batch. Note that `steps` refers to drawing one data point every hour, so steps is the number of hours in a sample; since we are looking back five days this dimension will always be five days worth of hourly observations.

(please forgive the formatting, my browser is having an issue)
`for j, row in enumerate(rows):` 
`            indices = range(rows[j] - lookback, rows[j], step)` 
`            samples[j] = data[indices] `
`            targets[j] = data[rows[j] + delay][1]` 
`        yield samples, targets`

^^ These lines of code actually build our samples.  We iterate through the list `rows` we created earlier which is a shuffled bunch of non-contiguous numbers (that we will use as index values).  `rows[j] - lookback` starts one observation (one sample), and `rows[j]` ends one observation.  We take all rows at `step` increments between these two values (in our case, `step` = 6).  So even though the rows we are taking were shuffled, the samples we build based on the numbers stored in `rows` are always five contiguous days sampled at ten minute intervals, with a target value 24 hours into the future.  Thus, having shuffled values in `rows` keeps the samples from being highly correlated and should help with training.

To drive the last point home, try running this code after the cell where you defined the generator:

`pretend_batch = np.zeros((10, 5, 14))`
^^ This will create a pretend batch matrix with `batch_size` = 10, `steps = 5`, `features = 14`

`float_data[[0,6,12,18,24]]`
^^ This gets five different rows of data, each one an hour apart (six steps of ten minutes each)

`pretend_batch[0] = float_data[[0,6,12,18,24]]`
^^ This sets the first sample in the pretend batch to be equal to those indices we looked at above

`pretend_batch`
^^ Notice how we have one sample at the top with the rest of the samples still being all zeros.


Hopefully this is clear and helpful.  I know it's been a couple of months since you posted this so you may have already figured this out by now, but maybe someone with the same question will find it helpful.

**Comment by jondo** [comment]
To summarize:
* Each sample contains the (unshuffled) lookback timeseries (see `data[indices]`).
* Shuffling takes the endpoints of these timeseries from random locations within the data.
